src/pieces.py

```diff
@@ -9,118 +9,272 @@
-class Knight(Piece):
-    """A knight."""
+class Knight(Melee):
+    """A knight.
+
+    Moves two squares in any direction and one square in a vertical to that direction as one move.
+    It therefore can jump over other pieces. Target square must not be blocked by a same side piece.
+    """
+
+#   Knight moves
+    steps = Melee.steps.union(diagonal + straight for diagonal, straight in product(Piece.diagonals, Piece.straights))
+    steps = steps.difference(Piece.rizontals.union(Piece.straights))  # remove the non-knight moves
 
     value: int = 3
 
     def __repr__(self) -> str:
-        """Represent the knight.
-
-        Returns:
-            The representation of a knight.
-        """
+        super().__repr__.__doc__
         return "♞" if self.is_black else "♘"
 
 
 @dataclass
-class Rook(Piece):
-    """A rook."""
+class Range(Piece):
+    """A long range piece:
+        - Rook
+        - Bishop
+        - Queen
+    """
+
+    def legal_moves(self, square: Square, condition: Callable[[Square], bool]) -> set[Square]:
+        super().legal_moves.__doc__
+        squares = set()
+        for move in self.steps:  # For each direction.
+            advanced_square = square + move  # Start looking at advanced squares
+            while condition(advanced_square):  # While we don't hit something.
+                squares.add(advanced_square)  # Add the damn square to legal destination squares.
+                advanced_square += move  # Advance to the next square in line.
+        return squares
+
+
+@dataclass
+class Rook(Range):
+    """A rook.
+
+    Moves anywhere along its rank xor its file, unless blocked by enemy (which captures) or same piece.
+    """
+
+#   Straight lines:
+    steps = Range.steps.union(Piece.straights)
 
     value: int = 5
 
     def __repr__(self) -> str:
-        """Represent the rook.
-
-        Returns:
-            The representation of a rook.
-        """
+        super().__repr__.__doc__
         return "♜" if self.is_black else "♖"
 
 
 @dataclass
-class Queen(Piece):
-    """A queen."""
+class Bishop(Range):
+    """A bishop.
 
-    value: int = 9
+    Moves anywhere along a diagonal, unless blocked by enemy (which captures) or same piece.
+    This means that a bishop never leaves the color of the square it is on.
+    A player starts with two bishops:
+        A black-square bishop.
+        A white-square bishop.
+    So both cover the squares of the whole board, and it is the reason the worth more when both on the board.
+    """
 
-    def __repr__(self) -> str:
-        """Represent the queen.
+#   Diagonal lines:
+    steps = Range.steps.union(Piece.diagonals)
 
-        Returns:
-            The representation of a queen.
-        """
-        return "♛" if self.is_black else "♕"
+    value: int = 3
+
+    def __repr__(self) -> str:
+        super().__repr__.__doc__
+        return "♝" if self.is_black else "♗"
 
 
 @dataclass
-class King(Piece):
-    """A king."""
+class Queen(Range, King):
```

Comment:
:warning: 
This is a smart observation but It is probably best to avoid multiple inheritance when possible.
`Queen` inherits from `Range` and `King` while `King` inherits from `Melee` which has a different `legal_moves` but gets overridden by `Range`.
This is too much puzzle work for someone reading this piece of code. It can be avoided with a single line:
```py
steps = Piece.straights | Piece.diagonals
```

I would favor the inheritance of `King` if it required more lines of code but here it is much more readable to explicitly state what the steps are. After all a big chunk of work in this PR (and multiple others) has been to be able to define this attribute in a clear and elegant way such as this one-liner.

This also fits better into Queen's legal moves conceptually. When reading from a chess textbook, it isn't based on the movements from the King.

Reply:
Done.